Channel list viewer

kangz12345 opened the pull request:
Implemented features:
- Refresh button: when clicked, fetch channels information from the backend server.
- ChannelList table: shows available channels and usage status.
- Use buttons: In the Use column, there are OX buttons and they are just toggling itself for now.
- Local storage: The channel list data is stored in `localStorage`, so refreshing the page will preserve the list state.

This closes #4.

kangz12345 committed:
Add channel types and slice

kangz12345 committed:
Register channelListReducer

kangz12345 committed:
Add channel list div and refresh button

kangz12345 committed:
Add ChannelList container

kangz12345 committed:
Add missing import

kangz12345 committed:
Implement thunk `fetch`

kangz12345 committed:
Implement reducer for fetch

kangz12345 committed:
Add ChannelList in MainPage

kangz12345 committed:
Fix warning about key prop

kangz12345 committed:
Modify indentations for readablity

kangz12345 committed:
Implement `toggleUse` action and export actions

kangz12345 committed:
Implement use toggling

kangz12345 committed:
Put empty lines for readability

kangz12345 committed:
Add useEffect hook to store channel list to local

kangz12345 committed:
Remove unnecessary arguments

kangz12345 committed:
Add trailing empty lines

kangz12345 committed:
Import react

kangz12345 committed:
Add trailing semicolons

kangz12345 committed:
Add a trailing semicolon

kangz12345 committed:
Add trailing semicolons and a comma

kangz12345 committed:
Add trailing semicolons

kangz12345 committed:
Fix eslint warning; use lower case boolean

kangz12345 committed:
Add a trailing semicolon

BECATRUE commented on wlm-ui/src/store/slices/channel/channel.ts:
```diff
@@ -0,0 +1,67 @@
+};
+
+const initialState: ChannelListInfo = {
+    channels: JSON.parse(localStorage.getItem('channel.channelList') ?? '[]'),
```
I used '||' before, however, '??' looks better👍
I will modify it later.

BECATRUE commented on wlm-ui/src/store/slices/channel/channel.ts:
```diff
@@ -0,0 +1,67 @@
+    channels: JSON.parse(localStorage.getItem('channel.channelList') ?? '[]'),
+};
+
+export const fetch = createAsyncThunk(
```
How about naming it something more detailed, like `fetchList`? It looks little vague.

BECATRUE commented on wlm-ui/src/store/slices/channel/channel.ts:
```diff
@@ -0,0 +1,67 @@
+            .addCase(fetch.fulfilled, (state, action) => {
+                const channels = action.payload.map((ch) => ch.channel);
+                const duplicates = state.channels
+                    .filter((info) => channels.includes(info.channel.channel));
```
I think it would neglect channels whose only name is updated.

BECATRUE commented on wlm-ui/src/store/slices/channel/channel.ts:
```diff
@@ -0,0 +1,67 @@
+                    ).map((ch) => <ChannelInfo>({ channel: ch, inUse: false }));
+                state.channels = duplicates.concat(newInfos).sort(
+                    (a, b) => a.channel.channel - b.channel.channel
+                );
```
Your implementation is correct, but it looks little complicated for me.

How about this?
```ts
state.channels = action.payload.map((ch) => {
    const originalInfo = state.channels.find((info) => info.channel.channel === ch.channel);
    return <ChannelInfo>({ channel: ch, inUse: (originalInfo ? originalInfo.inUse : false) });
}).sort((a, b) => a.channel.channel - b.channel.channel);
```
Or
```ts
let originalInfo;
state.channels = action.payload.map((ch) => <ChannelInfo>({
    channel: ch,
    inUse: (
        originalInfo = state.channels.find((info) => info.channel.channel === ch.channel)) ?
        originalInfo.inUse :
        false
    ),
})).sort((a, b) => a.channel.channel - b.channel.channel);
```

BECATRUE commented on wlm-ui/src/store/slices/channel/channel.ts:
```diff
@@ -0,0 +1,67 @@
+    reducers: {
+        toggleUse: (state, action: PayloadAction<{ channel: number }>) => {
+            const info = state.channels.find(
+                (chinfo) => chinfo.channel.channel === action.payload.channel
```
In below, you used an iteration variable named `info`, so how about unifying to one?

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import { useEffect } from "react";
```
Please replace double quotes with single ones.
There are several double quotes in the codes.

BECATRUE replied:
In addition, you can import this with the upper line, as below.
```ts
import React, { useEffect } from 'react';
```

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,42 @@
+import { useEffect } from "react";
+import { useDispatch, useSelector } from "react-redux";
+
+import { channelListActions, selectChannelList } from "../store/slices/channel/channel";
```
How about import a specific action directly, not this?

kangz12345 replied:
I used the reducer (not a thunk), so I think I cannot import each action.
Or maybe I can export the individual action in `channel.ts`?

BECATRUE replied:
Oh, I mistakenly thought you could import it directly.
Please leave it as it is.

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,42 @@
+
+import { channelListActions, selectChannelList } from "../store/slices/channel/channel";
+
+export default function ChannelListTable() {
```
It is just a convection, but how about defining this as a `const` value and exporting it at the end of the file?
Please refer to `user.ts`.

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,42 @@
+        return () => {
+            dispatch(channelListActions.toggleUse({ channel: channel }));
+        };
+    };
```
I'll leave this up to your taste. Which one looks better?
```ts
const onClickUse = (channel: number) => () => {
    dispatch(channelListActions.toggleUse({ channel: channel }));
};
```

kangz12345 replied:
!!! I actually like your suggestion 😄
Quite a cool way to do currying.

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,42 @@
+                    </tr>
+                );
+            })}</tbody>
+        </table>
```
We need to determine the HTML convection. For example,
- Attributes like `onClick` on the same line as the component name.
- Multiple components on the same line, like `<thead><tr>`.

kangz12345 replied:
I think too many line breaking is not good, because it makes too many indentations.
So my suggestion is:
- Try to write a tag e.g., `<button ...>...</button>`, `<input .../>`, in a single line if it is possible.
- If a tag has only a single tag inside, try to write it also in a single line, e.g., `<td><button ...>...</button></td>`.
- If a line break is required because of line length, try not to break a tag itself e.g., `<button ...>`, but its content: `<button ...>(\n)...(\n)</button>`.

Of course feel free to suggest any other idea!

BECATRUE replied:
I find it cool😃 Let's try to follow your suggestion.

kangz12345 committed:
Use single quotes instead of double quotes

kangz12345 committed:
Import in a single line

kangz12345 committed:
Avoid using `function`

kangz12345 commented on wlm-ui/src/MainPage/MainPage.tsx:
```diff
@@ -11,13 +13,26 @@ const MainPage = () => {
             </button>
+            <div>
+                <h1>Channel list</h1>
+                <button
+                    onClick={onClickRefreshChannelList}
+                >
```
Following my suggestion, this should be packed in a single line.

BECATRUE replied:
Then, please apply it!

kangz12345 committed:
Change `fetch` -> `fetchList` for clarity

kangz12345 committed:
Refactor updating the channel list, fixing a bug

Previous implementation did not handle the case where only the channel
name is changed.
The code suggested by Jaehun You.

Co-authored-by: Jaehun You <jaehun0121@snu.ac.kr>

kangz12345 committed:
Use consistent name for `ChannelInfo` objects

kangz12345 committed:
Use simpler form for currying

BECATRUE commented on wlm-ui/src/store/index.ts:
```diff
@@ -1,10 +1,12 @@
 
-import userReducer from "./slices/user/user";
+import channelListReducer from './slices/channel/channel';
+import userReducer from './slices/user/user';
```
Did you intend replacing two lines?

BECATRUE replied:
Oh.. You replaced the double quotes with single ones! Please never mind.

BECATRUE commented on wlm-ui/src/MainPage/ChannelList.tsx:
```diff
@@ -0,0 +1,41 @@
+
+    const dispatch = useDispatch();
+
+    const onClickUse = (channel: number) =>  () => {
```
There are two spaces in `=>  ()`!

kangz12345 committed:
Remove an unnecessary space

kangz12345 committed:
Remove unnecessary line breaks
